solidity/contracts/isms/block-hash/BlockHashIsm.sol

```diff
@@ -0,0 +1,104 @@
+    // strategy: using a block hash available through the oracle, the caller must provide a proof that the dispatch emit was emitted in that block.
+    // an efficient approach might try to use the block's event bloom filter.
+    // However, I avoid this because of the chance of false positives.
+    // This can be taken advantage of by attackers by finding events that populate the desired bits in a bloom filter.
+    function verify_internal(BlockHashMessageProof memory proof, bytes calldata message) internal view returns (bool) {
+        bytes32 blockHash = keccak256(proof.rlpBlockHeader);
+
+        // Verify that the block header provided is a block hash available from the oracle.
+        if (oracle.blockHash(proof.blockHeight) != uint256(blockHash)) return false;
+        // Verify that the origin of the message is the oracles' origin.
+        if (Message.origin(message) != ORIGIN) return false;
+
+        // Verify that the given tx receipt is included in the block header
+        if (
+            !MerklePatriciaProof.verify(
+                proof.rlpTxReceipt,
+                RLPEncode.encodeUint(proof.txIndex),
+                proof.txReceiptProof,
+                getReceiptRoot(proof.rlpBlockHeader)
+            )
+        ) return false;
+
+        // Verify that a dispatch event log with the given message was included in the tx receipt
+        if (!verifyMessageInReceipt(proof.rlpTxReceipt, message)) return false;
+
+        return true;
+    }
+
+    function getReceiptRoot(bytes memory rlpBlockHeader) internal pure returns (bytes32) {
+        RLPReader.RLPItem[] memory items = rlpBlockHeader.toRlpItem().toList();
+        return bytes32(items[5].toBytes());
+    }
+
+    function verifyMessageInReceipt(bytes memory rlpTxReceipt, bytes calldata message) internal view returns (bool) {
+        RLPReader.RLPItem[] memory logs = rlpTxReceipt.toRlpItem().toList()[3].toList();
+
+        // If we knew deterministically what log index the event would be, we could save gas by not iterating through all logs. I do not assume this here.
+        for (uint256 i; i < logs.length; i++) {
+            RLPReader.RLPItem[] memory logEntry = logs[i].toList();
+
+            // check emitter
+            if (address(uint160(logEntry[0].toAddress())) == MAILBOX) {
+                // check topics
+                RLPReader.RLPItem[] memory topics = logEntry[1].toList();
+                if (
+                    topics.length == 4 && bytes32(topics[0].toUint()) == DISPATCH_EVENT_SIGNATURE
+                        && address(uint160(topics[1].toAddress())) == Message.senderAddress(message)
+                        && uint32(topics[2].toUint()) == Message.destination(message)
+                        && bytes32(topics[3].toUint()) == Message.recipient(message)
+                        && keccak256(logEntry[2].toBytes()) == keccak256(message) // data field
```

Comment:
the `DispatchId` event might be easier to prove against

Reply:
Hmm, but how can we prove that the relayer did not manipulate the message contents in that case?

Reply:
Ah, I didn't notice that the message ID is actually the hash of the message. So yes, your suggestion is definitely the better way.

Reply:
true. I did not notice that the message ID is actually the message hash, so your solution is definitely better.